snuba/query/processors/prewhere.py

```diff
@@ -25,56 +34,189 @@ class PrewhereProcessor(QueryProcessor):
+        """
+        raise NotImplementedError
+
+
+class LegacyPrewhereProcessor(PrewhereProcessorDelegate[Condition, str]):
+    def _get_prewhere_candidates(
+        self, query: Query, prewhere_keys: Sequence[str]
+    ) -> Sequence[Tuple[Iterable[str], Condition]]:
         # Add any condition to PREWHERE if:
         # - It is a single top-level condition (not OR-nested), and
         # - Any of its referenced columns are in prewhere_keys
         conditions = query.get_conditions()
         if not conditions:
-            return
-        prewhere_candidates = [
+            return []
+        return [
             (util.columns_in_expr(cond[0]), cond)
             for cond in conditions
             if util.is_condition(cond)
             and cond[1] in ALLOWED_OPERATORS
             and any(col in prewhere_keys for col in util.columns_in_expr(cond[0]))
         ]
-        # Use the condition that has the highest priority (based on the
-        # position of its columns in the prewhere keys list)
-        prewhere_candidates = sorted(
+
+    def _get_column_name(self, column: str) -> str:
+        return column
+
+    def _update_conditions(
+        self, query: Query, prewhere_conditions: Sequence[Condition]
+    ) -> None:
+        conditions = query.get_conditions()
+        if not conditions:
+            # This should never happen at this point, but for mypy this can be None.
+            return
+
+        query.set_conditions(
+            [cond for cond in conditions if cond not in prewhere_conditions]
+        )
+        query.set_prewhere(prewhere_conditions)
+
+
+class ASTPrewhereProcessor(PrewhereProcessorDelegate[Expression, Column]):
+    """
+    This class MUST not depend on anything from the legacy implementation because,
+    as long as we have both ast and legacy representation coexisting, this class
+    cannot be sure whether the legacy representation has already been modified
+    to process the pre-where conditions.
+    """
+
+    allowed_ast_operators = [OPERATOR_TO_FUNCTION[o] for o in ALLOWED_OPERATORS]
+
+    def _get_prewhere_candidates(
+        self, query: Query, prewhere_keys: Sequence[str]
+    ) -> Sequence[Tuple[Iterable[Column], Expression]]:
+        # Add any condition to PREWHERE if:
+        # - It is a single top-level condition (not OR-nested), and
+        # - Any of its referenced columns are in prewhere_keys
+        ast_condition = query.get_condition_from_ast()
+        return (
             [
-                (
-                    min(
-                        prewhere_keys.index(col) for col in cols if col in prewhere_keys
-                    ),
-                    cond,
+                (get_columns_in_expression(cond), cond)
+                for cond in get_first_level_conditions(ast_condition)
+                if isinstance(cond, FunctionCall)
+                and cond.function_name in self.allowed_ast_operators
+                and any(
```

Comment:
I know you didn't actually change this, but shouldn't this be `all` not `any`? Does the prewhere still work if only a subset of the columns are in the prewhere key?

Reply:
I am not 100% sure what the reasoning was to pick prewhere condition if at least one of the columns included in the condition is a prewhere candidate vs all of them.
In practice the only case where we may have multiple column is if the condition we are looking into, on the left hand side has a function of multiple columns. 
At that point, I am not really sure the prewhere is actually very effective.